android/engine/src/test/java/org/smartregister/fhircore/engine/util/extension/StringExtensionTest.kt

```diff
@@ -38,4 +38,23 @@ class StringExtensionTest {
     Assert.assertEquals(expected, beforeFormatExampleOne.removeExtraWhiteSpaces())
     Assert.assertEquals(expected, beforeFormatExampleTwo.removeExtraWhiteSpaces())
   }
+
+  @Test
+  fun stringInterpolateShouldReplaceStrings() {
+    val templateString = "{ \"saveFamilyButtonText\" : @{ family.button.save } }"
+    val lookupMap = mapOf<String, Any>("family.button.save" to "Save Family")
+
+    Assert.assertEquals(
+      "{ \"saveFamilyButtonText\" : Save Family }",
+      templateString.interpolate(lookupMap)
+    )
+  }
+
+  @Test
+  fun stringInterpolateShouldThrowIllegalArgumentExceptionIfPrefixNull() {
+    val templateString = "{ \"saveFamilyButtonText\" : @{ family.button.save }, @{ missing } }"
+    val lookupMap = mapOf<String, Any>("family.button.save" to "save")
+
+    Assert.assertEquals(templateString, templateString.interpolate(lookupMap))
```

Comment:
this fails, because it's not generating an IllegalArgumentException, or any Exception. If it was it would return the unchanged string and the test would pass, but it's returning the changed string and the test is failing.

I'm not sure what the scenario is where we will generate the exception. Because the StringSubstitor has the flag the throw an IllegalArgumentException when it cannot find a string set to false.

What's the bug situation we're trying to account for with the change to `interpolate`?

Reply:
The exception we should be checking for is `IllegalStateException`. Additionally the exception occurs for an interesting edge case caused by the content of the key-value lookup map data.

I've updated the exception handling and unit test case to align with the above

Reply:
awesome thanks, but this test is not testing throwing the exception, if you comment out the try the throw doesn't happen. The idea with the map on line 56 in the test was that if exception gets thrown the interpolation should not happen. Can you update the test so that the exception throw is triggered and the result is tested?

Reply:
If we want to test the throwing of the exception directly then we need to remove the try catch.

Right now it is being tested implicitly because but I suppose there's no way of telling that from the test case data

Reply:
If we want to test the throwing of the exception directly then we need to remove the try catch.

Right now it is being tested implicitly but I suppose there's no way of telling that from the test case data

Reply:
I've updated the test case data to make it more apparent

Reply:
oh yea this is doing that, I consider this explicit :D looks great